liblokinet updates

majestrate opened the pull request:
required for liblokinet wrappers

majestrate committed:
update header with notes

majestrate committed:
update liblokinet udp api header

majestrate committed:
remove dead shit from header

majestrate committed:
update liblokinet udp header

majestrate committed:
changes for liblokinet-ffi

* cmake option BUILD_DAEMON for toggling building of daemon directory
* when WITH_BOOTSTRAP is OFF dont build curl or cpr

majestrate committed:
fix compile error

majestrate committed:
lokinet_add_bootstrap_rc

* allow bootstrap lists to be passed in

majestrate committed:
llarp::BootstrapConfig update

* make routers member a llarp::BootstrapList

majestrate committed:
typo fix

majestrate committed:
log errors on decoding

majestrate committed:
more logging

majestrate committed:
sanity check

majestrate committed:
more logging

majestrate committed:
correct function names

majestrate committed:
add WITH_BOOTSTRAP option for toggling building lokinet-bootstrap

majestrate committed:
typo fixes and clarify docs

majestrate committed:
fix typos add lokinet_udp_close

majestrate committed:
typofix

majestrate committed:
typofix

majestrate committed:
add lokinet_hex_to_base32z

majestrate committed:
wire up liblokient_udp_*

majestrate committed:
thread safety stuff

majestrate committed:
make it compile

majestrate committed:
move WITH_BOOTSTRAP option to root project CMakeLists.txt

majestrate committed:
enable liblokinet on macos

majestrate committed:
build liblokinet on macos

majestrate committed:
disable gost in static build

majestrate committed:
nodejs dipshittery

majestrate committed:
add liblokinet custom logger

majestrate committed:
dont enable apple languages when not building daemon

jagerman commented on contrib/mac.sh:
```diff
@@ -21,7 +21,7 @@ cmake \
       -DBUILD_SHARED_LIBS=OFF \
       -DBUILD_TESTING=OFF \
-      -DBUILD_LIBLOKINET=OFF \
+      -DBUILD_LIBLOKINET=ON \
```
Probably was only for testing, revert?

majestrate replied:
fixed

jagerman commented on external/CMakeLists.txt:
```diff
@@ -56,7 +56,7 @@ add_ngtcp2_lib()
 # 3.15+, and we target lower than that (and this is fairly simple to build).
-
+if(WITH_BOOTSTRAP)
 if(NOT BUILD_STATIC_DEPS)
```
Indent?

majestrate replied:
done

majestrate committed:
Update mac.sh

disable liblokinet on mac by default in mac.sh

majestrate committed:
Update CMakeLists.txt

dont depend on lokinet-shared target for sign target as we disabled it by default.

majestrate committed:
dont pack struct becuase alignment

jagerman commented on include/lokinet/lokinet_udp.h:
```diff
@@ -2,103 +2,119 @@
+      void* /*user*/,
+      const struct lokinet_udp_flowinfo* /* remote address */,
+      void** /* flow-userdata */,
+      int* /* timeout seconds */);
```
You can have argument names in a function pointer typedef, so these can all get uncommented.

majestrate replied:
names added

majestrate committed:
make function pointer arguments named

jagerman commented on include/lokinet/lokinet_udp.h:
```diff
@@ -2,103 +2,119 @@
-  };
+  /// @brief send on an established flow to remote endpoint
+  ///
+  /// @param flowinfo populated after call on success
```
This doc looks wrong

majestrate replied:
it is, fixed.

majestrate committed:
docstring update

majestrate committed:
identify

jagerman commented on llarp/lokinet_shared.cpp:
```diff
@@ -523,6 +824,13 @@ extern "C"
+  lokinet_hex_to_base32z(const char* hex)
+  {
+    const auto base32z = oxenmq::to_base32z(oxenmq::from_hex(std::string{hex}));
+    return strdup(base32z.c_str());
```
I'm thinking this should return nullptr if you give non-hex.  Also we can cut it down to a single allocation instead of 4 (3 strings + strdup) via:

```C++
char* EXPORT
lokinet_hex_to_base32z(const char* hex)
{
  std::string_view hexview{hex};
  if (!oxenmq::is_hex(hexview))
    return nullptr;

  const size_t byte_len = hexview.size() / 2;
  const size_t b32z_len = (byte_len*8+4) / 5; // = ⌈N×8÷5⌉ because 5 bits per 32z char
  auto buf = std::make_unique<char[]>(b32z_len + 1);
  char* end = buf.get() + b32z_len;
  *end = 0; // null terminate
  // Write the bytes into the *end* of the buffer so that when we rewrite the final b32z chars into
  // the buffer we won't overwrite any byte values until after we've consumed them.
  char* bytepos = end - byte_len;
  oxenmq::from_hex(hexview.begin(), hexview.end(), bytepos);
  // In-place conversion into the buffer
  oxenmq::to_base32z(bytepos, end, buf.get());
  return buf.release(); // leak the buffer to the caller
}
```

majestrate replied:
added

majestrate committed:
updates:

* add udptest example
* fix up udp codepath in liblokinet

majestrate committed:
cmake fixups

majestrate committed:
make pybind compile

majestrate committed:
typofix

tewinget commented on external/CMakeLists.txt:
```diff
@@ -56,26 +56,27 @@ add_ngtcp2_lib()
 # cpr configuration.  Ideally we'd just do this via add_subdirectory, but cpr's cmake requires
 # 3.15+, and we target lower than that (and this is fairly simple to build).
+if(WITH_BOOTSTRAP)
+  if(NOT BUILD_STATIC_DEPS)
```
this could be a single `if` line with the line above.

tewinget replied:
nvm misread.

tewinget commented on llarp/lokinet_shared.cpp:
```diff
@@ -28,6 +59,166 @@ namespace
+      bool isNewFlow{false};
+      {
+        std::unique_lock lock{m_Access};
+        isNewFlow = m_Flows.count(from) == 0;
```
if not `isNewFlow`, call HandlePacket here and return

majestrate replied:
revamped

tewinget commented on llarp/lokinet_shared.cpp:
```diff
@@ -28,6 +59,166 @@ namespace
+        // got a new flow, let's check if we want it
+        if (m_Filter(m_User, &flow_addr, &flow_userdata, &flow_timeoutseconds))
+          return;
+        AddFlow(from, flow_addr, flow_userdata, flow_timeoutseconds);
```
maybe add an optional argument to `AddFlow` so it can take a packet if we intend to send one right away after adding the flow.  This way we don't have to release the mutex acquired in AddFlow, then re-lock it; if we do have to re-lock it we should also re-check if the flow exists.

tewinget commented on llarp/service/endpoint.hpp:
```diff
@@ -168,6 +170,12 @@ namespace llarp
       HandlePathDied(path::Path_ptr p) override;
 
+      virtual vpn::EgresPacketRouter*
+      EgresPacketRouter()
```
I love this speling of egres.

tewinget commented on llarp/vpn/packet_router.hpp:
```diff
@@ -17,7 +17,6 @@ namespace llarp::vpn
 
     virtual void AddSubHandler(nuint16_t, PacketHandlerFunc){};
   };
-
```
style nitpick: empty lines between class declarations are nice to have.

majestrate committed:
make it compile

majestrate committed:
resolve race condition in udp flow and packet handling

majestrate committed:
do less allocations in lokinet_hex_tobase32z
